Mathlib/Combinatorics/Hypergraph/Basic.lean

```diff
@@ -0,0 +1,492 @@
+(unordered) sets of vertices; i.e., they are subsets of the vertex set `V`.
+
+A hypergraph where `V = ∅` and `E = ∅` is *empty*. A hypergraph with a nonempty
+vertex set (`V ≠ ∅`) and empty hyperedge set is *trivial*.
+
+If a hyperedge `e` contains only one vertex (i.e., `|e| = 1`), then it is a *loop*.
+
+This module defines `Hypergraph α` for a vertex type `α` (hyperedges are defined as `Set (Set α)`).
+
+## Main definitions
+
+For `H : Hypergraph α`:
+
+* `V(H)` denotes the vertex set of `H` as a term in `Set α`.
+* `E(H)` denotes the hyperedge set of `H` as a term in `Set (Set α)`. Hyperedges must be subsets of
+    `V(H)`.
+* `H.Adj x y` means that there exists some hyperedge containing both `x` and `y` (or, in other
+    words, `x` and `y` are incident on some shared hyperedge `e`).
+* `H.EAdj e f` means that there exists some vertex that is incident on both hyperedge `e` and
+    hyperedge `f : Set α`.
+
+## Implementation details
+
+This implementation is heavily inspired by Peter Nelson and Jun Kwon's `Graph` implementation,
+which was in turn inspired by `Matroid`.
+
+Paraphrasing `Mathlib.Combinatorics.Graph.Basic`:
+"The main tradeoff is that parts of the API will need to care about whether a term
+`x : α` or `e : Set α` is a 'real' vertex or edge of the graph, rather than something outside
+the vertex or edge set. This is an issue, but is likely amenable to automation."
+
+Because `hyperedgeSet` is a `Set (Set α)`, rather than a multiset, here we are assuming that
+all hypergraphs are *without repeated hyperedge*.
+
+## Acknowledgments
+
+Credit to Shreyas Srinivas, GitHub user @NotWearingPants ("Snir" on the Lean Zulip), Ammar
+Husain, Aaron Liu, and Tristan Figueroa-Reid for patient guidance and useful feedback on this
+implementation.
+-/
+
+open Set
+
+variable {α β γ : Type*} {x y : α} {e e' f g : Set α} {l : Set (Set α)}
+
+/--
+An undirected hypergraph with vertices of type `α` and hyperedges of type `Set α`,
+as described by vertex and hyperedge sets `vertexSet : Set α` and `hyperedgeSet : Set (Set α)`.
+
+The requirement `hyperedge_isSubset_vertexSet` ensures that all vertices in hyperedges are part of
+`vertexSet`, i.e., all hyperedges are subsets of the `vertexSet`.
+-/
+@[ext]
+structure Hypergraph (α : Type*) where
+  /-- The vertex set -/
+  vertexSet : Set α
+  /-- The hyperedge set -/
+  hyperedgeSet : Set (Set α)
+  /-- All hyperedges must be subsets of the vertex set -/
+  hyperedge_isSubset_vertexSet' : ∀ ⦃e⦄, e ∈ hyperedgeSet → e ⊆ vertexSet
+
+namespace Hypergraph
+
+variable {H : Hypergraph α}
+
+/-! ## Notation -/
+
+/-- `V(H)` denotes the `vertexSet` of a hypergraph `H` -/
+scoped notation "V(" H ")" => Hypergraph.vertexSet H
+
+/-- `E(H)` denotes the `hyperedgeSet` of a hypergraph `H` -/
+scoped notation "E(" H ")" => Hypergraph.hyperedgeSet H
+
+
+instance : Membership (α) (Hypergraph α) where
+  mem H x := x ∈ V(H)
+
+instance : Membership (Set α) (Hypergraph α) where
+  mem H e := e ∈ E(H)
```

Comment:
I think we should leave these out for now, they're already not used (I think) in this PR, and I'm not yet convinced that they're a good idea (despite me writing the second one!)